[PbD Interim Fix]: Update EligibilityWarnings layout

p-jackson opened the pull request:
## Changes proposed in this Pull Request

Fixes the Private by Default upgrade experience as desribed here paObgF-I3-p2

This PR tries to focus on the layout changes of the `<EligibilityWarnings>` component. Let me know if there's something you think could be split into a seperate PR.

* Removes the "upgrade to business" `<Banner>` and moves it into the list of other holds
* Restyle the list of holds to use checkmarks
* Restyle the list of warnings to look consistent with holds
* Dim and disable the main warnings card when there's a blocking `<Notice>` at the top e.g. when you get a `BLOCKED_ATOMIC_TRANSFER` error

*Before*
<img width="730" alt="Screenshot 2019-12-12 at 5 46 37 PM" src="https://user-images.githubusercontent.com/1500769/70683504-7b5eac00-1d07-11ea-9884-7f2a3b8def6a.png">

*After*
See testing instructions

## Testing instructions

### On /plugins/upload

#### Free, unlaunched/private site should see:

<img width="740" alt="Zrzut ekranu 2019-12-20 o 12 40 42" src="https://user-images.githubusercontent.com/205419/71252703-ffc3b700-2325-11ea-9bb8-ceb02c538d25.png">

#### Free, public site should see:

<img width="742" alt="Zrzut ekranu 2019-12-20 o 12 40 46" src="https://user-images.githubusercontent.com/205419/71252722-0a7e4c00-2326-11ea-8c46-9e838e7709f0.png">

#### Business plan, unlaunched/private site should see:

<img width="741" alt="Zrzut ekranu 2019-12-20 o 12 40 53" src="https://user-images.githubusercontent.com/205419/71252709-04886b00-2326-11ea-8a86-d6dfed864e44.png">

#### Business plan, public site should see:

<img width="758" alt="Zrzut ekranu 2019-12-20 o 13 07 29" src="https://user-images.githubusercontent.com/205419/71253996-cdb45400-2329-11ea-82b5-297dea8b1f45.png">

#### Unlaunched site with a paid domain should see:

<img width="744" alt="Zrzut ekranu 2019-12-20 o 12 43 50" src="https://user-images.githubusercontent.com/205419/71252856-65b03e80-2326-11ea-9ad4-6de73224379b.png">

#### Business plan, public site, with a paid domain should see:

<img width="739" alt="Zrzut ekranu 2019-12-20 o 13 12 13" src="https://user-images.githubusercontent.com/205419/71254163-64811080-232a-11ea-9715-01b6f55d57b6.png">

### On `/themes/upload`

Repeat the same process as above. The only difference should be the header and the copy. It should talk about themes instead of plugins.

### On `/hosting-config/activate`

Repeat the same process as above. The only difference should be the header and the copy. It should talk about activating hosting instead of plugins.

### With blog sticker

Add the `blocked-from-atomic-transfer` sticker to your site, you should see the following:

<img width="740" alt="Zrzut ekranu 2019-12-20 o 12 59 24" src="https://user-images.githubusercontent.com/205419/71253569-9d1fea80-2328-11ea-97c5-529b9d6d7e01.png">

### Other considerations

* Click on Continue button, confirm a tracks event `cta_name=calypso-theme-eligibility-upgrade-nudge` is still getting logged
* Try the same test plan with a mobile device - is everything still looking good there?
* After switching UI to another language, fallback copy should be used:

<img width="741" alt="Zrzut ekranu 2019-12-20 o 15 34 40" src="https://user-images.githubusercontent.com/205419/71261729-69e85600-233e-11ea-85a9-ea7cfeef0803.png">

## Known issues

* The copy in this PR seems a little bit weird, it effectively says "You'll need to: Public site". This is going to be addressed in another PR that will be released in tandem with this one: https://github.com/Automattic/wp-calypso/pull/38324
* We still have a "Help" button that links to support pages and is effectively a dead end. This is in line with what's currently on that screen and will be addressed in another PR.

Fixes #38141

p-jackson committed:
Add business plan error to hold list

Also removed code for NOT_USING_CUSTOM_DOMAIN which is no longer
returned by the backend.

p-jackson committed:
Add business plan error to hold list

Also removed code for NOT_USING_CUSTOM_DOMAIN which is no longer
returned by the backend.

p-jackson committed:
Restyle hold list to use checkmarks

p-jackson committed:
Restyle hold list to use checkmarks

p-jackson committed:
AT blockers appear at the top of <EligibilityWarnings>

p-jackson committed:
AT blockers appear at the top of <EligibilityWarnings>

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -94,24 +65,82 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		SITE_GRAYLISTED: {
+			message: translate(
+				"There's an ongoing site dispute. Contact us to review your site's standing and resolve the dispute."
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Contact us to review your site's standing and resolve the dispute.' )` *ES Score: 6*

jblz replied:
Merging the titles and messages like this is going to result in a number of places where we'll be waiting for localized strings.  Should we either:
* String freeze this PR & hold for translations
* Do something like: `message: hasTranslation( "There's an ongoing site dispute. Contact us to review your site's standing and resolve the dispute." ) ? translate( "There's an ongoing site dispute. Contact us to review your site's standing and resolve the dispute." : translate( 'Contact us to review your site's standing and resolve the dispute.' ),`

adamziel replied:
I was thinking about shipping this one without waiting for translations since it's a way better experience than what we currently have, `hasTranslation` is just perfect - I'm going to update this PR right now.

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -94,24 +65,82 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		NO_SSL_CERTIFICATE: {
+			message: translate(
+				'Certificate installation in progress. Hold tight! We are setting up a digital certificate to allow secure browing on your site using "HTTPS".'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Hold tight! We are setting up a digital certificate to allow secure browsing on your site, using "HTTPS". Please try again in a few minutes.' )` *ES Score: 6*

p-jackson committed:
Dim the fixable holds when a blocking hold is visible

p-jackson committed:
Dim the fixable holds when a blocking hold is visible

p-jackson committed:
Remove business upgrade banner from <EligibilityWarnings>

We're showing that a business plan is required in the holds list for
consistency inside this component.

Removing banner allows some mocking to be removed from tests.

p-jackson committed:
Remove business upgrade banner from <EligibilityWarnings>

We're showing that a business plan is required in the holds list for
consistency inside this component.

Removing banner allows some mocking to be removed from tests.

p-jackson committed:
Add aria-label prop to <NoticeAction>

This allows actions with only an icon to add a label for screen readers.

p-jackson committed:
Add aria-label prop to <NoticeAction>

This allows actions with only an icon to add a label for screen readers.

p-jackson committed:
Move AT warnings to top of UI, seperate from the holds area

p-jackson committed:
Move AT warnings to top of UI, seperate from the holds area

p-jackson committed:
Remove unused styles

Styles unused now that warnings use <Notice>

p-jackson committed:
Remove unused styles

Styles unused now that warnings use <Notice>

p-jackson committed:
Refactor <EligibilityWarnings> to define the <Card> outline

p-jackson committed:
Refactor <EligibilityWarnings> to define the <Card> outline

p-jackson committed:
Increase space between heading and hold list

p-jackson committed:
Restyle eligibility warnings confirmation button

p-jackson committed:
Restyle eligibility warnings confirmation button

p-jackson committed:
Go to checkout from <EligibilityWarnings> if user has no blocking holds

p-jackson committed:
Go to checkout from <EligibilityWarnings> if user has no blocking holds

p-jackson committed:
Fix test that wasn't testing anything

The default window.location in Jest is example.com, so my test wasn't
doing anything. Doesn't look like jsdom actually follows links so the
best way to test a link will do the right thing is just to test its
href.

p-jackson committed:
Fix test that wasn't testing anything

The default window.location in Jest is example.com, so my test wasn't
doing anything. Doesn't look like jsdom actually follows links so the
best way to test a link will do the right thing is just to test its
href.

ramonjd commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +65,91 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		TRANSFER_ALREADY_EXISTS: {
+			message: translate(
+				'Installation in progress. Just a minute! Please wait until the installation is finisehd, then try again.'
```
`finisehd`

ramonjd commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +65,91 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		NO_SSL_CERTIFICATE: {
+			message: translate(
+				'Certificate installation in progress. Hold tight! We are setting up a digital certificate to allow secure browing on your site using "HTTPS".'
```
`browing`

ramonjd commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -151,16 +190,37 @@ export const HoldList = ( { holds, isPlaceholder, translate }: Props ) => {
+function getCardHeading( context: string | null, translate: LocalizeProps[ 'translate' ] ) {
+	switch ( context ) {
+		case 'plugins':
+			return translate( "To install plugins, you'll need a couple things:" );
```
`couple things`

ramonjd commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -151,16 +190,37 @@ export const HoldList = ( { holds, isPlaceholder, translate }: Props ) => {
+		case 'plugins':
+			return translate( "To install plugins, you'll need a couple things:" );
+		case 'themes':
+			return translate( "To install themes, you'll need a couple things:" );
```
as above: missing `of`

ramonjd commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -151,16 +190,37 @@ export const HoldList = ( { holds, isPlaceholder, translate }: Props ) => {
+		case 'themes':
+			return translate( "To install themes, you'll need a couple things:" );
+		default:
+			return translate( "To continue, you'll need a couple things:" );
```
as above: missing `of`

p-jackson replied:
I was not typing well that day it seems! 😅 

The "couple things" came from the design. I don't mind it, but do you feel it's too informal compared to our usual voice?

ramonjd replied:
I've heard it before, and shudder every time :)

I know it's part of North-American English, but in my judgement, readers from the UK/AU/NZ/CA would find it unusual. Confusing things further, "couple things" is also used to describe activities that take place between couples in a relationship.

We're speaking to a wider audience, and therefore I'd feel comfortable overriding the copy writers on this one, if only to make the sentence clearer, and easier to read.

Just my 1.39c (after AUD->USD conversion)

ramonjd reviewed (changes requested):
I've tested according to instructions, and the screens look good

<img width="300" alt="Screen Shot 2019-12-12 at 4 45 30 pm" src="https://user-images.githubusercontent.com/6458278/70686426-70ece400-1d00-11ea-83a7-15d61a4fde33.png">
<img width="300" alt="Screen Shot 2019-12-12 at 4 45 43 pm" src="https://user-images.githubusercontent.com/6458278/70686427-70ece400-1d00-11ea-938a-aad4d40b90b4.png">
<img width="300" alt="Screen Shot 2019-12-12 at 4 48 01 pm" src="https://user-images.githubusercontent.com/6458278/70686428-71857a80-1d00-11ea-9603-26967ec97b80.png">

Not confident on commenting on the code being a bit out of the loop.

Requesting changes as there are some typos in the copy.

p-jackson committed:
typos

p-jackson committed:
typos

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +65,91 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		TRANSFER_ALREADY_EXISTS: {
+			message: translate(
+				'Installation in progress. Just a minute! Please wait until the installation is finished, then try again.'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Just a minute! Please wait until the installation is finished, then try again.' )` *ES Score: 8*

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +65,91 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+		},
+		NO_SSL_CERTIFICATE: {
+			message: translate(
+				'Certificate installation in progress. Hold tight! We are setting up a digital certificate to allow secure browsing on your site using "HTTPS".'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Hold tight! We are setting up a digital certificate to allow secure browsing on your site, using "HTTPS". Please try again in a few minutes.' )` *ES Score: 8*

adamziel commented:
Relevant discussion going on at paObgF-Mn-p2

adamziel committed:
Implement Ian's updates to the design

adamziel committed:
Implement Ian's updates to the design

adamziel committed:
Update copy

adamziel committed:
Update copy

adamziel committed:
Remove a comma from warnings title

adamziel committed:
Remove a comma from warnings title

p-jackson committed:
Fixup tests after changes to <WarningList>

p-jackson committed:
Fixup tests after changes to <WarningList>

p-jackson committed:
Copy for when component appears in /themes and /hosting-config contexts

p-jackson committed:
Copy for when component appears in /themes and /hosting-config contexts

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +70,91 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+	return {
+		BLOCKED_ATOMIC_TRANSFER: {
+			message: translate(
+				'This site is not currently eligible to install themes and plugins, or activate hosting access. Please contact our support team for help.'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'This site is not currently eligible to install themes and plugins. Please contact our support team for help.' )` *ES Score: 7*

p-jackson commented:
Updated some copy that assumed `<EligibilityWarnings>` was going to appear during plugin uploading. But it's also used during theme upload and when configuring hsoting options https://github.com/Automattic/wp-calypso/pull/38277/commits/c9547b077e5c20c83b4fca2685f1f4c6c0a2fd3f

When describing the hosting I said things like: `These features aren't (yet) compatible with`**`hosting access`**`and will be disabled` and `This site is not currently eligible to install themes and plugins, or`**`activate hosting access`**.

I got nomenclature from the notice on `/hosting-config`:
<img width="959" alt="Screenshot 2019-12-20 at 4 43 16 PM" src="https://user-images.githubusercontent.com/1500769/71229663-51326d00-234b-11ea-9687-0fb59811dcd3.png">

adamziel committed:
Remove icons for warnings list

adamziel committed:
Remove icons for warnings list

adamziel commented:
Thank you so much @p-jackson, this is super thoughtful!

I just implemented the rest of design updates and this is ready for review again. CC @michaeldcain @ianstewart

adamziel committed:
Update CSS classes to set entire warnings/holds list into indented/unindented mode (vs specific messages)

adamziel committed:
Update CSS classes to set entire warnings/holds list into indented/unindented mode (vs specific messages)

jblz committed:
JSDoc fixes to appease the linter

jblz commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -151,16 +194,39 @@ export const HoldList = ( { holds, isPlaceholder, translate }: Props ) => {
+		case 'hosting':
+			return translate( "To activate hosting access you'll need to:" );
+		default:
+			return translate( "To continue you'll need to" + ':' );
```
```suggestion
			return translate( "To continue you'll need to:" );
```

jblz commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -1,44 +1,35 @@
 import { localizeUrl } from 'lib/i18n-utils';
 
 // Mapping eligibility holds to messages that will be shown to the user
 // TODO: update supportUrls and maybe create similar mapping for warnings
```
Is this TODO still relevant after the change?

jblz commented on client/blocks/eligibility-warnings/index.tsx:
```diff
@@ -38,66 +34,34 @@ interface ExternalProps {
-		);
-	}
+	const logEventAndProceed = () => {
+		// We removed <Banner /> Component which logged this event on upsell click.
```
I'm not sure this comment is useful in the codebase going forward.  I'd document it in the PR / commit history and take out the comment

jblz commented on client/blocks/eligibility-warnings/test/index.tsx:
```diff
@@ -0,0 +1,172 @@
+import React, { ReactChild } from 'react';
+import { createStore } from 'redux';
+import { Provider } from 'react-redux';
+import { fireEvent, render } from '@testing-library/react';
```
😍

jblz reviewed (commented):
The code looks really good overall -- Great work! I'm continuing to test.

I added a concern about string changes that we should probably solve before landing this.

Also, I think it would be helpful to add is more developer documentation around eligibility warnings in general and about specific functions.

Can we please either / and / or:
* Add a `client/blocks/eligibility-warnings/README.md` which describes the purpose of the component at a high level
* Add JSDoc blocks to functions like:
  * `isHardBlockingHoldType`
  * `getBlockingMessages`

...so unfamiliar folks (& future us) will have an easier time groking what this does and how it relates to the data from the API

jblz commented:
Noting that the `ci/wp-desktop` failure looks related to the node upgrade.

jblz commented on client/blocks/eligibility-warnings/index.tsx:
```diff
@@ -38,66 +34,34 @@ interface ExternalProps {
+		// We removed <Banner /> Component which logged this event on upsell click.
+		// We want to keep tracking the data in the same way, so let's fake the banner click here
+		recordCtaClick( feature );
+		onProceed();
```
It looks like the component is not actually blocking the plugin flow (at least):

![eligibility-warning-doesnt-block](https://user-images.githubusercontent.com/1587282/71259607-d3e50900-2306-11ea-96a1-effca3e8eee3.gif)

Do we still need to build out the implementation for the click handler to launch the site, add the plan to the cart, & redirect to checkout?

adamziel replied:
That's listed in known issues

adamziel replied:
See https://github.com/Automattic/wp-calypso/pull/38364

jblz replied:
Cool.  That's all I've found -- I'll go review that PR now so we can get it merged in here.

adamziel replied:
This is now part of this PR so it should be fully functional

adamziel committed:
hasTranslation fallbacks

adamziel committed:
hasTranslation fallbacks

adamziel committed:
Address nitpicks

adamziel committed:
Address nitpicks

adamziel committed:
Add docstrings

adamziel committed:
Add docstrings

adamziel committed:
Handle proceed button

adamziel committed:
Handle proceed button

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -93,24 +87,115 @@ function getHoldMessages( translate: LocalizeProps[ 'translate' ] ) {
+				'Certificate installation in progress. Hold tight! We are setting up a digital certificate to allow secure browsing on your site using "HTTPS".'
+			)
+				? hasTranslation(
+						'Certificate installation in progress. Hold tight! We are setting up a digital certificate to allow secure browsing on your site using "HTTPS".'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Hold tight! We are setting up a digital certificate to allow secure browsing on your site, using "HTTPS". Please try again in a few minutes.' )` *ES Score: 9*

jblz commented on client/blocks/eligibility-warnings/hold-list.tsx:
```diff
@@ -151,16 +235,57 @@ export const HoldList = ( { holds, isPlaceholder, translate }: Props ) => {
 };
 
+function getCardHeading( context: string | null, translate: LocalizeProps[ 'translate' ] ) {
+	const defaultCopy = translate( 'Please clear all issues above to proceed.' );
```
Unfortunately, this copy (`Please clear all issues above to proceed.`) doesn't work since the issues now appear below the Heading text.  I'll just update it to the new fallback: `To continue you'll need to:` & hopefully the individual items will be enough context for non-English speakers.

jblz replied:
This should be fixed in 3a756ef

jblz committed:
Use fallback text for header vs the incorrect "above" string

jblz reviewed (changes requested):
Varying the text by consulting `i18n.hasTranslation` was not working properly in the default (US English) locale.  Fix in #38562

jblz commented on client/blocks/eligibility-warnings/test/index.tsx:
```diff
@@ -0,0 +1,181 @@
+		fireEvent.click( upgradeAndContinue );
+
+		expect( handleProceed ).not.toHaveBeenCalled();
+		expect( upgradeAndContinue ).toHaveAttribute(
```
This is the only test failing once the strings are fixed by #38562.

I think we need to do something like:
* Attach an `onClick` to the component which:
  * Dispatches the action to launch the site, then
  * Redirects to the checkout screen
* Update this test to cover the above

adamziel replied:
That test is off, this isn’t supposed to work in this PR. For now the button is just greyed out. There is a separate PR to add „auto-launching” feature. It’s okay to remove this test from here.

adamziel replied:
Ah I got confused by `Dispatches the action to launch the site, then` - I took a deeper look and this test should pass. Your follow-up PR will fix it though as the button will have the correct copy (https://github.com/Automattic/wp-calypso/pull/38562).

adamziel committed:
Make hasTranslation aware of default locale

adamziel committed:
Use custom hasLocalizedText function to check if new copy may be used

adamziel commented:
As `i18n-calypso` patch may not be straightforward to merge (See https://github.com/Automattic/wp-calypso/pull/38565 and https://github.com/Automattic/wp-calypso/pull/38566), I added a local version of `hasLocalizedText()` function what we may use for now. Let's remove it once the other patch gets merged or this one gets properly translated.

adamziel committed:
Make hasLocalizedText the default import

adamziel committed:
Adjust test to page.redirect instead of href

jblz reviewed (commented):
This is working really well.  Once this is rebased w/ unrelated changes removed, I think it's good to go!

adamziel committed:
Add newline at the end of the has-localized-text

adamziel commented on client/blocks/eligibility-warnings/has-localized-text.js:
```diff
@@ -0,0 +1,9 @@
+export const hasLocalizedText = message =>
+	i18n.state.localeSlug === i18n.defaultLocaleSlug || i18n.hasTranslation( message );
+
+export default hasLocalizedText;
```
A linter rules would be a good follow up to this

jblz reviewed (approved):
Bravo!

adamziel commented:
This one was quite a journey - deployed now :-)
